src/x/debugdump/heap.go

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package debugdump
+
+import (
+	"io"
+	"runtime/pprof"
+)
+
+// heapDumpProvider provides a heap dump of the process
+type heapDumpProvider struct{}
```

Comment:
How do you feel about adding a CPU provider as well? That would make this P.R significantly more useful once its landed.

I think it should be as simple as something like:

```golang
type cpuProfileProvider struct {
    duration time.Duration
}

func (c *cpuProfileProvider) ProvideData(w io.Writer) error {
    err := pprof.StartCPUProfile(w); err != nil {
        return err
    }

    time.Sleep(h.duration)
    pprof.StopCPUProfile()
}

Ideally we'd be able to control the duration at runtime too, so maybe add a sync.Mutex into the struct and then a public method for modifying the duration, but you get the idea.

Reply:
Yeah, makes perfect sense. Though I'm not sure about controlling the duration, you'd have to keep the object somewhere, because the ZipWriter doesn't expose the Sources.
Or it could have a `Source(filename string) Source`.